tfe/resource_tfe_workspace_run.go

```diff
@@ -0,0 +1,157 @@
+package tfe
+
+import (
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+var (
+	backoffMin = 1000.0
+	backoffMax = 3000.0
+)
+
+var planPendingStatuses = map[tfe.RunStatus]bool{
+	tfe.RunPending:        true,
+	tfe.RunPlanQueued:     true,
+	tfe.RunPlanning:       true,
+	tfe.RunCostEstimating: true,
+	tfe.RunPolicyChecking: true,
+	tfe.RunQueuing:        true,
+	tfe.RunFetching:       true,
+}
+
+var planTerminalStatuses = map[tfe.RunStatus]bool{
+	tfe.RunPlanned:            true,
+	tfe.RunPlannedAndFinished: true,
+	tfe.RunErrored:            true,
+	tfe.RunCostEstimated:      true,
+	tfe.RunPolicyChecked:      true,
+	tfe.RunPolicySoftFailed:   true,
+	tfe.RunPolicyOverride:     true,
+}
+
+var applyPendingStatuses = map[tfe.RunStatus]bool{
+	tfe.RunConfirmed:   true,
+	tfe.RunApplyQueued: true,
+	tfe.RunApplying:    true,
+	tfe.RunQueuing:     true,
+	tfe.RunFetching:    true,
+}
+
+var applyDoneStatuses = map[tfe.RunStatus]bool{
+	tfe.RunApplied: true,
+	tfe.RunErrored: true,
+}
+
+var confirmationDoneStatuses = map[tfe.RunStatus]bool{
+	tfe.RunConfirmed:   true,
+	tfe.RunApplyQueued: true,
+	tfe.RunApplying:    true,
+}
+
+func resourceTFEWorkspaceRun() *schema.Resource {
+	return &schema.Resource{
+		Create:        resourceTFEWorkspaceRunCreate,
+		Delete:        resourceTFEWorkspaceRunDelete,
+		Read:          resourceTFEWorkspaceRunRead,
+		Update:        resourceTFEWorkspaceRunUpdate,
+		SchemaVersion: 1,
+		Schema: map[string]*schema.Schema{
+			"workspace_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"apply": {
+				Type:         schema.TypeList,
+				Elem:         resourceTFEWorkspaceRunSchema(),
+				Optional:     true,
+				AtLeastOneOf: []string{"apply", "destroy"},
+				MaxItems:     1,
+			},
+			"destroy": {
+				Type:     schema.TypeList,
+				Elem:     resourceTFEWorkspaceRunSchema(),
+				Optional: true,
+				MaxItems: 1,
+			},
+		},
+	}
+}
+
+func resourceTFEWorkspaceRunCreate(d *schema.ResourceData, meta interface{}) error {
+	// var isDestroyRun & currentRetryAttempts is declared for the sole purpose of code readability
+	isDestroyRun := false
+	currentRetryAttempts := 0
+	return createWorkspaceRun(d, meta, isDestroyRun, currentRetryAttempts)
+}
+
+func resourceTFEWorkspaceRunDelete(d *schema.ResourceData, meta interface{}) error {
+	// var isDestroyRun & currentRetryAttempts is declared for the sole purpose of code readability
+	isDestroyRun := true
+	currentRetryAttempts := 0
+	return createWorkspaceRun(d, meta, isDestroyRun, currentRetryAttempts)
+}
+
+func resourceTFEWorkspaceRunUpdate(d *schema.ResourceData, meta interface{}) error {
+	// update is a noop since this resource only creates a run during a destroy or an initial apply phase
+	return nil
+}
+
+func resourceTFEWorkspaceRunRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
```

Comment:
Here's where the problem was: if a run has no apply (and hence a random ID), the Read fails with a 404. If that causes us to `SetID("")`, then the resource gets deleted from the state, and during a Terraform destroy, that means we decide the resource is already gone and hence there's nothing to do during the destruction! It's absent during the destroy plan, and you don't get a workspace destroy run. 

I think the solution is to just bail early if we know there shouldn't be any run to read. That leaves the ID (and hence resource) in place, so we get our proper destroy on schedule. This patch worked for me. 

```suggestion
	// First check whether this is a destroy-only run
	_, ok := d.GetOk("apply")
	if !ok {
		// If there's no apply, then there won't be anything to "read" until we
		// do a destroy run. Return now and leave the ID alone, so that we keep
		// the resource in the state and get a destroy run when the time comes.
		log.Printf("[DEBUG] Run %s (random ID) has no apply; nothing to read for refresh", d.Id())
		return nil
	}

	config := meta.(ConfiguredClient)
```